### PR TITLE
SSD cache-dictionary

### PR DESCRIPTION
**Pull request by nikvas0**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Cache-dictionaries layout for storing data on SSD.

TODO:
- [x] getBlockInputStream
- [x] Counters
- [x] Strings support
- [x] one file fifo compaction
- [x] lru for keys in memory
- [x] Configurable write buffer size
- [x] Checksums for blocks
- [x] store keys and metadata on ssd?
- [x] Any keys support
- [x] Get rid of duplicate objects in update from different threads
- [x] Perfomance test
- [x] Change HashMap (now unordered_map) (store hashes instead of keys?)
- [x] documentation
- [x] some refactoring

### CODE REVIEW

**Comment by filimonov** [comment]
Reinventing aerospike (non distributed) ? :) 
See #2718 #5421

**Comment by azat** [comment]
>Reinventing aerospike (non distributed) ? :)

rocksdb?

**Comment by nikvas0** [comment]
> > Reinventing aerospike (non distributed) ? :)
> 
> rocksdb?

I think I should also write rocksdb-based implementation and compare with this dictionary.

Maybe RocksDB with FIFO compaction + lightweight compactions + disabled WAL will be a good solution. Also, there is a table format (https://github.com/facebook/rocksdb/wiki/PlainTable-Format) with in-memory index.

**Comment by nikitamikhaylov** [comment]
I think it is necessary to add performance tests to benchmark it in comparison with DirectDictionary and CacheDictionary. Or just  CacheDictionary in memory with SSDCacheDictionary. 
I don't mean performance tests in CI, I want something like the one given here #10622

File: src/Dictionaries/SSDCacheDictionary.cpp [hunk 1]
```diff
@@ -0,0 +1,1698 @@
...
+    PaddedPODArray<Key> array;
+    for (const auto & key : key_to_index.keys())
+        array.push_back(key);
+    return array;
+}
+
+void SSDCachePartition::remove()
+{
+    std::unique_lock lock(rw_lock);
+    std::filesystem::remove(std::filesystem::path(path + BIN_FILE_EXT));
```
Comment (nikitamikhaylov) [inline src/Dictionaries/SSDCacheDictionary.cpp:902]: Error handling?

File: src/Dictionaries/SSDCacheDictionary.cpp [hunk 2]
```diff
@@ -0,0 +1,1698 @@
...
+
+    std::unordered_map<Key, std::vector<size_t>> not_found_ids;
+    storage.getValue<OutputType>(attribute_index, ids, out, not_found_ids, get_default, now);
+    if (not_found_ids.empty())
+        return;
+
+    std::vector<Key> required_ids(not_found_ids.size());
+    std::transform(std::begin(not_found_ids), std::end(not_found_ids), std::begin(required_ids), [](const auto & pair) { return pair.first; });
+
+    storage.update(
```
Comment (nikitamikhaylov) [inline src/Dictionaries/SSDCacheDictionary.cpp:1413]: Maybe better to wrap it in a separate function, which adds source_ptr and lifetime and call this function?

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Dictionaries/BucketCache.h [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                return bucket * bucket_size + cur;
+            }
+        }
+
+        return bucket * bucket_size + pos;
+    }
+
+    size_t getPosition(const size_t bucket) const
+    {
+        const size_t idx = (bucket >> 1);
```
[the remark this answers is not in the document]
Comment (nikvas0) [reply src/Dictionaries/BucketCache.h:180]: To determine a position in the bucket we need to read 4 bits from the array with positions, but we can address only bytes, so we need to divide bucket number by 2.

**Comment by filimonov** [comment]
BTW - it would be really interesting to see some perf number, and/or comparison with aerospike (#5629)

**Comment by alexey-milovidov** [comment]
@filimonov They are published in https://presentations.clickhouse.tech/hse_2020/3rd/SSD_Dictionary_full.pdf

**Comment by filimonov** [comment]
Just for the reference - aerospike gives up to ~1Mln req/sec

**Comment by alexey-milovidov** [comment]
@filimonov This number does not mean anything.
It can be too slow or too fast depending on hardware.
Nikita has tested with laptop SSD.

**Comment by nikitamikhaylov** [comment]
Merged in #11947

**Pull request closed without merging**